### ISSUE TITLE
Following imports

### ISSUE THREAD

**Issue by LightAndLight** [qfpl/hpython#27]
Eventually I want to be able to parse and validate imported modules. Not going to do it yet, but I'm going to do some brainstorming here.

https://docs.python.org/3.5/reference/import.html#searching

* We'll have to duplicate the "finders and loaders" logic
* Should probably make import awareness work for calls to `importlib.import_module` as well as import statements
* Need to warn about importing inside control flow, as we can't give useful guarantees

**Comment by LightAndLight** [comment]
Yep that's the 3rd bullet point :) I've come up against similar issues in scope checking. I wish it wasn't this way.

**Comment by andreabedini** [comment]
> Yep that's the 3rd bullet point :)

😊 

> I've come up against similar issues in scope checking. I wish it wasn't this way.

Reading `https://docs.python.org/3.5/library/functions.html#__import__` is a big depressing.

> This function is invoked by the import statement. It can be replaced (by importing the builtins module and assigning to builtins.__import__) in order to change semantics of the import statement, but doing so is strongly discouraged as it is usually simpler to use import hooks (see PEP 302) to attain the same goals and does not cause issues with code which assumes the default import implementation is in use. Direct use of __import__() is also discouraged in favor of importlib.import_module().

I wonder what guarantees can be given about Python at all.

**Comment by tonymorris** [comment]
> I wonder what guarantees can be given about Python at all.

The `+` function does not format your disk.

**Comment by tonymorris** [comment]
maybe, hopefully

**Comment by andreabedini** [comment]
```
>>> int.__add__ = better_add
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: can't set attributes of built-in/extension type 'int'
```

At this point I am not sure if I should be happy or sad ...

**Comment by LightAndLight** [comment]
Definitely happy, because I've tried that same thing before :P
